app/controllers/invoices_controller.rb

```diff
@@ -1,7 +1,144 @@
+    else
+      @invoice.status = 'draft'
+    end
+    if @invoice.save
+      if params[:email_send] == "true"
+        pdf_render
+        ClientMailer.send_email(@invoice.client,@kit.to_pdf, @invoice, current_user).deliver
+      end
+      flash[:success] = "Invoice created successfully."
+      redirect_to invoices_path
+    else
+      flash[:error] = "Invoice not saved because: #{add_flash_messages(@invoice)}"
+      render :new
+    end
+  end
+
+  def update
+    if @invoice.update(invoice_params)
+      if @invoice.address.present?
+        @invoice.address.tap do |a|
+          invoice_address = @invoice.client.address
+          a.street_1 = invoice_address.street_1
+          a.street_2 = invoice_address.street_2
+          a.city = invoice_address.city
+          a.state = invoice_address.state
+          a.pincode = invoice_address.pincode
+          a.country_code = invoice_address.country_code
+        end
+        if @invoice.address.save
+          if params[:email_send] == "true"
+            @invoice.status = 'sent'
+          else
+            @invoice.status = 'draft'
+          end
+          if @invoice.save
+            if params[:email_send] == "true"
+              pdf_render
+              ClientMailer.send_email(@invoice.client, @kit.to_pdf, @invoice, current_user).deliver
+            end
+          flash[:success] = "Invoice updated successfully."
+          redirect_to invoices_path
+          end
+        else
+          flash[:error] = "Invoice not updated because: #{add_flash_messages(@invoice)}"
+          render :new
+        end
+      else
+          @invoice.address = Address.new.tap do |a|
+            invoice_address = @invoice.client.address
+            a.street_1 = invoice_address.street_1
+            a.street_2 = invoice_address.street_2
+            a.city = invoice_address.city
+            a.state = invoice_address.state
+            a.pincode = invoice_address.pincode
+            a.country_code = invoice_address.country_code
+          end
+          if params[:email_send] == "true"
+            @invoice.status = 'sent'
+          else
+            @invoice.status = 'draft'
+          end
+          if @invoice.save
+            if params[:email_send] == "true"
+              pdf_render
+              ClientMailer.send_email(@invoice.client, @kit.to_pdf, @invoice, current_user).deliver
+            end
+            flash[:success] = "Invoice updated successfully."
+            redirect_to invoices_path
+          end
+      end
+    else
+      flash[:error] = "Invoice not updated because: #{add_flash_messages(@invoice)}"
+      render :new
+    end
+  end
+
+  def edit
+    render :new
+  end
+
+  def destroy
+    if @invoice.destroy
+      #@invoice.address.destroy if @invoice.address.present?
+      flash[:success] = "Invoice deleted successfully."
+      redirect_to invoices_path
+    else
+      flash[:error] = "Invoice not deleted because: #{add_flash_messages(@invoice)}"
```

Comment:
`Unable to delete Invoice #{invoice_number}.  Reason - #{add_flash_messages(@invoice)}"